### DISCUSSION TITLE
Get all changes made to entities

### DISCUSSION THREAD

**Discussion by Langstra** [mikro-orm/mikro-orm#5207] [Ideas]
**Is your feature request related to a problem? Please describe.**
For B4nan a tl;dr: https://mikroorm.slack.com/archives/CFSAF8DEX/p1668519465260469   
I would like to determine the changes between before the flush and after the flush. This enables you to generate a change log/audit trail. Currently there are some options like `uow.getChangeSets()` and `uow.getCollectionUpdates()` that you can call in `onFlush()` and `afterFlush()` event handlers. In the AfterFlush event I tried to do uow.getCollectionUpdates() and then compare the collectionUpdate.getSnapshot() to the collectionUpdate.getItems(), but those are the same. Doing this in the OnFlushEvent does give a difference between the getSnapshot()and getItems() . Problem with the OnFlush is that while creating an entity and adding it to the collection like book.authors(em.create(Author, {})will then not give the pk (which is not strange since it is in the onflush). Also I realise that the getSnapshot() is an internal function, so I should probably not rely on it.

**Describe the solution you'd like**
During the flush event we could cache the changes during the onFlush event and check the pk's during the afterFlush event. Then expose all the changes in a separate method instead of using the internal methods that I am doing right now.

**Describe alternatives you've considered**
Considered doing this myself with the help of the asyncStorage.

**Comment by B4nan** [comment]
Just to add more context - in after flush hook you always get the state - well - after the flush. So yes, seeing the snapshots already updated is correct, expected, and entity states behave exactly the same (collection snapshot is equivalent to the original entity data) - that event is there to hook after the flush operation. 

I am not sure if I want to generate more in-memory data to slow down the flush operation just so you can read the changes in a hook that is not meant for it. I'd be ok with that only if it has no performance impact (it could be opt-in if there is some). So I don't see this as a problem with after flush event, rather with onFlush event, if we can make the changes available in after flush, it will be a pro, but not a necessity (you can always cache them yourself).

I think we could have some method on the ChangeSet class that would do it. It would still work only in onFlush, if we want to make it lazy, it cant work in after flush (without caching = early evaluation = possible perf degradation). Whats the reason why you want to use after flush?

**Comment by B4nan** [comment]
Maybe we should rethink this a bit, and try to make the collection diffs part of the change sets. IIRC the reason why they are handled separately is to not fire on update hooks for entities where only collection items changed - but people often want this, so maybe it would be the right thing to do. Although I guess it would end up as breaking change, so better keep such refactor to v6.

**Comment by Langstra** [comment]
>  Whats the reason why you want to use after flush?

Problem I was currently facing is that I could get the collection changes from inside the `onFlush` hook, but that for newly created entities in the collection there was no pk value. Which seems logic, since the entities are not flushed yet. So that is why I thought the `afterFlush` hook was more logical.

**Comment by B4nan** [comment]
Would be good to have some test case for this. I made few smaller changes in master so maybe the issue is no longer present, PK values should be no longer wiped I think, only nullable FKs will.

**Comment by B4nan** [comment]
I will move this to discussions as right now this is not actionable, I need some reproduction to see what is missing (and I need to be able to verify things on the current version). I know it was my idea, but that does not mean it was just an idea, and creating an issue out of it requires more work than just pasting a link (not to mention the link is no longer working as the content expired).
